### ISSUE TITLE
text render blend func may use BlendFuncSeparate better

### ISSUE THREAD

**Issue by riikack** [fyne-io/fyne#3923]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same feature, before opening a new one.
- [X] This issue only relates to a single feature. I will open new issues for any other features.

### Is your feature request related to a problem?

when set
glfw.WindowHint(glfw.TransparentFramebuffer, glfw.True)
i found text render has something wrong on its outline (alpha problem)

### Is it possible to construct a solution with the existing API?

no

### Describe the solution you'd like to see.

drawSingleChannelTexture
p.ctx.BlendFunc(srcAlpha, oneMinusSrcAlpha)
->
p.ctx.BlendFuncSeparate(srcAlpha, oneMinusSrcAlpha, one, oneMinusSrcAlpha)

**Comment by andydotxyz** [comment]
It would really help with a screenshot to understand what you mean.

**Comment by riikack** [comment]
![image](https://github.com/fyne-io/fyne/assets/32995902/79ff96a0-1776-4e51-82c7-09766db89002)
![image](https://github.com/fyne-io/fyne/assets/32995902/d318d0f1-0d8d-488e-a223-7436868111ba)
![image](https://github.com/fyne-io/fyne/assets/32995902/04ae2d20-2bb9-4707-a303-1d2cea76cbf5)

**Comment by andydotxyz** [comment]
I'm a little unclear about what white / black background you mean when it looks like a mid-grey.
Additionally I don't understand how this relates to the transparent window as the examples seem to show opaque rectangles.
Sorry for the questions, just trying to understand.

**Comment by riikack** [comment]
I mean, there is a potential rendering error here: the blend mode of text. Due to the semi transparent sampling of the texture edges of the text, according to the current rendering method, the semi transparent effect still penetrates into the background when there is a rectangle behind it. Obviously, this is not expected because blend should not be calculated together with color when calculating alpha, but should be calculated separately

The transparent window mode only more significantly exposes this issue and is not directly related to it

**Comment by andydotxyz** [comment]
Hmm, do you have a custom theme loaded by any chance?
If the colour is specified as `RGBA` then it may exhibit problems not present with `NRGBA` where the values are raw (not pre-multiplied).

Just trying to understand more, because we place text colour on various different colours and it hasn't been a problem before.

**Comment by riikack** [comment]
Yes, I found that your rendering was correct without setting `glfw.WindowHint (glfw. TransparentFramebuffer, glfw. True)`

However, I still recommend using BlendFuncSeparate instead of BlendFunc, as BlendFuncSeparate can render correctly in both cases. If you need to support transparent windows in the future, this change seems necessary

Sorry, I haven't tried rendering under normal circumstances before

**Comment by andydotxyz** [comment]
I don't understand how your screenshots were generated though - the background is not painted by the text, but in your transparent examples there were rectangles under all.
Also I don't understand why your proposed code solves it - moving the alpha channel to One->oneMinusSrcAlpha (away from SrcAlpha) does not change what colour is blended, it just ensures that the alpha is either the same as requested or more opaque (if the font colour had alpha < 0xff).

**Comment by riikack** [comment]
My screenshot was mainly achieved by modifying `glfw.WindowHint(glfw.TransparentFramebuffer, glfw.True)`
In addition, it is also necessary to comment out the `p.ctx.ClearColor(float32(r)/max16bit, float32(g)/max16bit, float32(b)/max16bit, float32(a)/max16bit)` in the painter


You have already given the answer to the second question. The problem with transparent window mode lies in alpha blend, where color blend is inherently correct

**Comment by andydotxyz** [comment]
Removing the clear colour doesn't sound right as the context should be clearer (avoiding random memory).
Instead you should set the theme
background to transparent. Let me know how it looks when you take that approach instead.

**Comment by riikack** [comment]
do you mean something like `(CustomTheme) Color(...)` -> `case theme.ColorNameBackground: return color.RGBA{R: 0, G: 0, B: 0}`
to set the theme background to transparent?

this cannot make the background transparent

**Comment by andydotxyz** [comment]
Yes that is what I mean. Of course you will also still need to tell GLFW to make it transparent as well or you will get random memory on the window I expect.

Also do try and use `NRGBA` as I suggested earlier - it's just a good habit to get in to, and often avoids unexpected alpha issues as most folk's heads do not work on pre-multiplied alpha maths ;).
